### PR TITLE
1-bit adam

### PR DESCRIPTION
**Pull request by awan-10**

### CODE REVIEW

File: deepspeed/runtime/fp16/onebit_adam.py [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+class OnebitAdam(torch.optim.Optimizer):
+    """Implements the 1-bit Adam algorithm. Currently GPU-only.
+    For usage example please see, TODO DeepSpeed Tutorial
+    It has been proposed in TODO: add paper name and arXiv link.
+
+    Arguments:
+        params (iterable): iterable of parameters to optimize or dicts defining
+            parameter groups.
+        lr (float, optional): learning rate. (default: 1e-3)
+        betas (Tuple[float, float], optional): coefficients used for computing
```
Comment (jeffra) [inline deepspeed/runtime/fp16/onebit_adam.py:27]: Update the docstring here to add details about all the arguments that OnebitAdam uses, e.g., freeze_step is a pretty important one to document here (and other places too).

File: deepspeed/runtime/fp16/onebit_adam.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+        compensated_server_m = compensated_server_m.add_(1).bool()
+        cupy_server_scale = self.torch2cupy(server_scale)
+        cupy_compensated_server_m = self.torch2cupy(compensated_server_m)
+        #print(cupy_compensated_server_m[0:10])
+        compensated_server_m = None
+        # del compensated_server_m
+
+        cupy_server_sign_packed = self.compress_by_chunk(cupy_compensated_server_m, 1)
+        #print('rank is : {}, the tensor before  allgather is {}'.format(rank, cupy_server_sign_packed[0][0:10]))
+        #print(cupy.unpackbits(cupy_server_sign_packed[0][0]))
```
Comment (jeffra) [inline deepspeed/runtime/fp16/onebit_adam.py:189]: Can we remove some of the debug comments throughout the files?

File: requirements/requirements.txt [hunk 3]
```diff
@@ -3,3 +3,5 @@ torchvision>=0.4.0
 tqdm
 psutil
 tensorboardX==1.8
+cupy-cuda100
+mpi4py
```
Comment (jeffra) [inline requirements/requirements.txt:7]: Let's move these into a specific requirements.txt for 1bit adam. I'll take care of this though.

File: deepspeed/runtime/fp16/onebit_adam.py (hunk 2 above)
Comment (awan-10) [reply deepspeed/runtime/fp16/onebit_adam.py]: Cleaned up.

File: deepspeed/runtime/fp16/onebit_adam.py (hunk 1 above)
Comment (awan-10) [reply deepspeed/runtime/fp16/onebit_adam.py:29]: Add the detail.

File: deepspeed/runtime/fp16/onebit_adam.py (hunk 2 above)
Comment (awan-10) [reply deepspeed/runtime/fp16/onebit_adam.py]: Done :)

File: deepspeed/runtime/fp16/onebit_adam.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+        if not self.initialize:
+            self.adam_freeze_key = False
+            self.initialize = True
+            print(
+                f"Finished the initialization step at rant {torch.distributed.get_rank()}"
+            )
+            return loss
+
+        if self.adam_freeze_key is False:
+            if state['step'] >= self.freeze_step:
```
Comment (niumanar) [inline deepspeed/runtime/fp16/onebit_adam.py:352]: Suggestion: could we try to set freeze_step empirically using a simple norm test on exp_avg between the baselineAdam and onebitAdam?

**Comment by ShadenSmith** [comment]
Marked PR as a draft until we are ready to merge.

File: deepspeed/runtime/fp16/onebit_adam.py [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+                        0.999),
+                 eps=1e-8,
+                 eps_inside_sqrt=False,
+                 weight_decay=0.,
+                 max_grad_norm=0.,
+                 amsgrad=False,
+                 cuda_aware=False):
+
+        if amsgrad:
+            raise RuntimeError('FusedLamb does not support the AMSGrad variant.')
```
Comment (jeffra) [inline deepspeed/runtime/fp16/onebit_adam.py:66]: FusedLamb -> OnebitAdam

File: deepspeed/runtime/fp16/onebit_adam.py [hunk 6]
```diff
@@ -0,0 +1,374 @@
+'''
+Copyright 2019 The Microsoft DeepSpeed Team
```
Comment (jeffra) [inline deepspeed/runtime/fp16/onebit_adam.py:2]: 2019 -> 2020

File: deepspeed/runtime/fp16/onebit_adam.py (hunk 4 above)
Comment (jeffra) [reply deepspeed/runtime/fp16/onebit_adam.py:370]: Let's defer this until after Wednesday. Sorry @niumanar, but I do think we should look into this after our release deadline!

File: deepspeed/runtime/utils.py [hunk 7]
```diff
@@ -32,6 +33,11 @@ def check_using_norm(self, norm_group):
                                          op=torch.distributed.ReduceOp.MAX,
                                          group=self.mpu.get_model_parallel_group())
             overflow = overflow_gpu[0].item()
+        else:
+            cuda_overflow = torch.cuda.FloatTensor([overflow])
+            dist.all_reduce(cuda_overflow, op=torch.distributed.ReduceOp.MAX)
+            dist.barrier()
+            overflow = cuda_overflow[0].item()
```
Comment (jeffra) [inline deepspeed/runtime/utils.py:40]: I don't think we need to do this. This function is used in fp16/fused_optimizer.py and fp16/unfused_optimizer.py as well as in 1bit-adam. If we really need to do this for 1bit-adam let's have an `elif` here and pass a parameter into `check_using_norm` to do the reduction. I believe in standard fp16 training with deepspeed at this point all ranks will have already all-reduced their gradients which means everyone will have an overflow or no one will have an overflow.

**Review by jeffra — APPROVED** [review]

File: deepspeed/runtime/utils.py (hunk 7 above)
Comment (awan-10) [reply deepspeed/runtime/utils.py:40]: Need to check with Hank on this one. I did not add these lines. I will update this when I hear back.

File: deepspeed/runtime/fp16/onebit_adam.py (hunk 6 above)
Comment (awan-10) [reply deepspeed/runtime/fp16/onebit_adam.py]: Done. I thought this will happen when we grep for all others that are 2019 as well :)

File: deepspeed/runtime/fp16/onebit_adam.py (hunk 5 above)
Comment (awan-10) [reply deepspeed/runtime/fp16/onebit_adam.py]: Done.

File: deepspeed/runtime/utils.py (hunk 7 above)
Comment (jeffra) [reply deepspeed/runtime/utils.py:40]: Sounds good, let's merge this into staging v2 as soon as we get this comment settled.
